vault/namespaces_store.go

```diff
@@ -0,0 +1,757 @@
+	core    *Core
+	storage logical.Storage
+
+	// This lock ensures we don't concurrently modify the store while using
+	// a namespace entry. We also store an atomic to check if we need to
+	// reload all namespaces.
+	lock        sync.RWMutex
+	invalidated atomic.Bool
+
+	// List of all namespaces within the store. This is loaded at store
+	// initialization time and persisted throughout the lifetime of the
+	// instance. Entries should not be returned directly but instead be
+	// copied to prevent modification.
+	namespacesByPath     *namespaceTree
+	namespacesByUUID     map[string]*namespace.Namespace
+	namespacesByAccessor map[string]*namespace.Namespace
+
+	// logger is the server logger copied over from core
+	logger hclog.Logger
+}
+
+// NewNamespaceStore creates a new NamespaceStore that is backed
+// using a given view. It used used to durable store and manage named namespace.
+func NewNamespaceStore(ctx context.Context, core *Core, logger hclog.Logger) (*NamespaceStore, error) {
+	ns := &NamespaceStore{
+		core:                 core,
+		storage:              core.barrier,
+		logger:               logger,
+		namespacesByPath:     newNamespaceTree(nil),
+		namespacesByUUID:     make(map[string]*namespace.Namespace),
+		namespacesByAccessor: make(map[string]*namespace.Namespace),
+	}
+
+	// Add namespaces from storage to our table. We can do this without
+	// holding a lock as we've not returned ns to anyone yet.
+	if err := ns.loadNamespacesLocked(ctx); err != nil {
+		return nil, fmt.Errorf("error loading initial namespaces: %w", err)
+	}
+
+	return ns, nil
+}
+
+func (ns *NamespaceStore) checkInvalidation(ctx context.Context) error {
+	if !ns.invalidated.Load() {
+		return nil
+	}
+
+	ns.lock.Lock()
+	defer ns.lock.Unlock()
+
+	// Status might have changed
+	if !ns.invalidated.Load() {
+		return nil
+	}
+
+	if err := ns.loadNamespacesLocked(ctx); err != nil {
+		return fmt.Errorf("error handling invalidation: %w", err)
+	}
+
+	ns.invalidated.Store(false)
+	return nil
+}
+
+// loadNamespaces loads all stored namespaces from disk. It assumes the lock
+// is held when required.
+func (ns *NamespaceStore) loadNamespacesLocked(ctx context.Context) error {
+	// Assume we roughly have as many namespaces as we have presently. During
+	// invalidation this will pre-allocate enough space to reload everything
+	// as we'll likely be essentially in sync already. However, at startup, this
+	// will mostly just give us space for the root namespace.
+	namespacesByPath := newNamespaceTree(namespace.RootNamespace)
+	namespacesByUUID := make(map[string]*namespace.Namespace, len(ns.namespacesByUUID)+1)
+	namespacesByAccessor := make(map[string]*namespace.Namespace, len(ns.namespacesByAccessor)+1)
+	namespacesByUUID[namespace.RootNamespaceUUID] = namespace.RootNamespace
+	namespacesByAccessor[namespace.RootNamespaceID] = namespace.RootNamespace
+
+	if err := logical.WithTransaction(ctx, ns.storage, func(s logical.Storage) error {
```

Comment:
I think we should actually split this so child namespaces' storage references are in the parent, not root. We need another recursion here, but that way if we have an external import of a namespace ([see vision](https://openbao.org/blog/vision-for-namespaces/)), you get children for free without having to explicitly register it. My 2c.

Reply:
I think this is necessary for out-of-hierarchy namespaces (parent-less namespaces), based on conversation with @schultz-is.